### PR TITLE
[core][experimental] Accelerated DAG: Infer torch.Tensor shape and metadata on first DAG execution

### LINKED ISSUE #45595: [core][experimental] Improve automatic shape and type declaration for torch.Tensors in accelerated DAG
### Description

For p2p GPU communication, we need to know the tensor metadata on the receiver in order to allocate a receiving buffer of the correct size. Sending tensor metadata to the receiver can add overhead and is unnecessary when the tensor shape and dtype is known ahead of time. The accelerated DAG API lets the user specify the shape and dtype ahead of time to avoid this overhead. However, the current API is clunky because:
- it requires the user to specify the shape and dtype at compile time
- it does not work for values that contain multiple tensors

Ideally, we should instead expose a single flag that says whether the shape(s) and dtype(s) will remain static, and then infer the metadata on the first result. This would resolve both of the above usability issues.

Previous API:
```python
dag = sender.send.bind()
dag = dag.with_type_hint(TorchTensorType(shape=(N, ), dtype=torch.float16))
receiver.recv.bind(dag)
```

Proposed new API:
```python
dag = sender.send.bind()
dag = dag.with_type_hint(TorchTensorType(static_shape=True))
receiver.recv.bind(dag)
```

### Use case

_No response_

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

For p2p GPU communication, we need to know the tensor metadata on the receiver in order to allocate a receiving buffer of the correct size. Sending tensor metadata to the receiver can add overhead and is unnecessary when the tensor shape and dtype is known ahead of time. The accelerated DAG API lets the user specify the shape and dtype ahead of time to avoid this overhead. Similarly, we have the flag `direct_return=True` to indicate when a task returns only GPU tensor data and no other data that must be transferred via host memory. However, the current API is clunky because:

- it requires the user to specify the shape and dtype at compile time
- it does not work for values that contain multiple tensors

Previous API:

```python
dag = sender.send.bind()
dag = dag.with_type_hint(TorchTensorType(_shape=(N, ), _dtype=torch.float16), _direct_return=True)
receiver.recv.bind(dag)
```

Proposed new API - the plan is to keep `_direct_return` for now but replace the shape and dtype args with `_static_shape` flag.

```python
dag = sender.send.bind()
dag = dag.with_type_hint(TorchTensorType(_static_shape=True, _direct_return=True))
receiver.recv.bind(dag)
```
In this PR, we now dynamically infer the torch.Tensor shape and dtype based on the first DAG execution, instead of requiring the former to be declared by the user.

In particular, we always allocate two shared memory channels per TorchTensorNcclChannel: one for the tensor(s) metadata and one for the serialized CPU data. If `static_shape=True`, we read the metadata channel once and reuse the value for subsequent messages. Else, we read from the metadata channel on each message.

## Related issue number

Closes #45595.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/experimental/channel/shared_memory_channel.py [hunk 1]
```diff
@@ -363,9 +337,19 @@ def write(self, value: Any):
             self._num_readers,
         )
 
-    def begin_read(self) -> Any:
+    def begin_read(self, deserialize=True) -> Any:
         self.ensure_registered_as_reader()
```
Comment (kevin85421) [inline python/ray/experimental/channel/shared_memory_channel.py:341]: Add comments

**Review by kevin85421 — COMMENTED** [review]
I am still reading the PR. I will continue reviewing it tomorrow.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 2]
```diff
@@ -50,31 +61,49 @@ def __init__(
...
+            readers: The actors that may read from the channel. None signifies
+                the driver.
+            tensor_data_channel: A GPU-GPU channel for sending tensor data. Its
+                writer and readers should match the given writer and readers.
+            non_tensor_data_channel: A shared-memory channel for sending
+                non-tensor data. Its writer and readers should match the given
+                writer and readers.
+            static_non_tensor_data: Whether the non-tensor data sent through
+                this channel is static. For example, if all written values have
+                the form `{"my_tensor": torch.Tensor(...)}`, then this can be
```
Comment (kevin85421) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:77]: I am a bit confused. Why is `{"my_tensor": torch.Tensor(...)}` considered non-tensor data?

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 3]
```diff
@@ -84,48 +113,59 @@ def __init__(
...
+            (
+                self._writer,
+                self._readers,
+                self._tensor_data_channel,
+                self._non_tensor_data_channel,
+                self._static_non_tensor_data,
+            ),
         )
 
     def ensure_registered_as_writer(self):
```
Comment (kevin85421) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:128]: Do we need to call this function in `def write`? Currently, we rely on the underlying channels (i.e., `_tensor_data_channel` and `_non_tensor_data_channel`'s `ensure_xxx` functions) being idempotent. Do we need to add `self._writer_registered` to ensure this function is always idempotent?

```python
def ensure_registered_as_writer(self):
        if self._writer_registered:
            return
        ...
        self._writer_registered = True
```

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 4]
```diff
@@ -84,48 +113,59 @@ def __init__(
...
-        self._gpu_data_channel.ensure_registered_as_writer()
-        if self._cpu_data_channel is not None:
-            self._cpu_data_channel.ensure_registered_as_writer()
+        self._tensor_data_channel.ensure_registered_as_writer()
+        self._non_tensor_data_channel.ensure_registered_as_writer()
 
     def ensure_registered_as_reader(self):
```
Comment (kevin85421) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:132]: Do we need to call the function in `def begin_read` and `def end_read`?

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 5]
```diff
@@ -1,45 +1,56 @@
...
     from ray.experimental.channel.shared_memory_channel import Channel
-    from ray.experimental.channel.torch_tensor_type import TorchTensorType
 
 
 # Logger for this module. It should be configured at the entry point
 # into the program using Ray. Ray provides a default configuration at
 # entry/init points.
 logger = logging.getLogger(__name__)
 
+# 100KB to store metadata and/or exceptions.
```
Comment (jackhumphries) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:27]: 100 kB

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 6]
```diff
@@ -50,31 +61,49 @@ def __init__(
         writes the worker-local
         ray.experimental.channel.serialization_context._SerializationContext
         when serializing data.
+
+        Args:
+            writer: The actor that may write to the channel. None signifies the
+                driver.
+            readers: The actors that may read from the channel. None signifies
+                the driver.
```
Comment (jackhumphries) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:69]: We should sync about this. I changed this part of the API in #45786, so we should discuss how to make them consistent.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 2 above)
Comment (jackhumphries) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:77]: I think it's because the tensors are within a dictionary, which is itself not a tensor. Since there are non-tensor components in the data, we can set `static_non_tensor_data=True` even though the dictionary contains tensors.

File: python/ray/dag/tests/experimental/test_mocked_nccl_dag.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+    # Sending correct shape still errors because the DAG has already been torn
+    # down after the previous error.
+    spec.append(("c", 1, shape, dtype))
+    with pytest.raises(OSError, match="Channel closed"):
+        output_channel = compiled_dag.execute(spec)
+
+    # TODO(swang): Ideally we would also check that sending a different value,
+    # e.g., changing the keys of the dictionary or sending a list instead of a
+    # dictionary, would also fail, but currently we don't have a reliable way
+    # to do this. Checking the serialized output produces many false positives.
```
Comment (jackhumphries) [inline python/ray/dag/tests/experimental/test_mocked_nccl_dag.py:374]: Can we check this in the `write()` function after calling `reset_out_of_band_tensors()`? Maybe we could zero out all of the tensor references and then take the checksum.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 8]
```diff
@@ -138,63 +178,130 @@ def write(self, value: Any):
...
+                )
+
+        # First send the extracted tensors through a GPU-specific channel.
+        self._tensor_data_channel.write(tensors_to_send)
+
+        # Next send the non-tensor data, if needed. The data contains
+        # placeholders for the extracted tensors, through a CPU-specific
+        # channel.
+        if self._num_serialized_tensors is None:
+            self._non_tensor_data_channel.write(serialized_non_tensor_data)
```
Comment (jackhumphries) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:213]: Is it guaranteed that this write will be visible to the other device after the prior write (`self._tensor_data_channel.write(tensors_to_send)`) is visible? How is this synchronized?

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 9]
```diff
@@ -138,63 +178,130 @@ def write(self, value: Any):
...
-        if self._cpu_data_channel:
-            self._cpu_data_channel.end_read()
+        self._tensor_data_channel.end_read()
+
+        if not self._static_non_tensor_data:
+            # If the non-tensor data is static, then we reuse the serialized
+            # data in the shared memory buffer for all future channel messages.
+            # Therefore, we should only release the buffer if the non-tensor
+            # data is dynamic.
+            self._non_tensor_data_channel.end_read()
```
Comment (ruisearch42) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:273]: This looks incompatible and will introduce complication for `CompiledDAGRef` API?

File: python/ray/experimental/channel/torch_tensor_type.py [hunk 10]
```diff
@@ -130,52 +131,41 @@ def create_channel(
         self,
         writer: Optional["ray.actor.ActorHandle"],
         readers: List[Optional["ray.actor.ActorHandle"]],
-        _torch_tensor_allocator: Optional["TorchTensorAllocator"] = None,
+        _non_tensor_data_channel: Optional["Channel"] = None,
+        _tensor_metadata_channel: Optional["Channel"] = None,
```
Comment (ruisearch42) [inline python/ray/experimental/channel/torch_tensor_type.py:135]: Add documentation what these are for

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 2 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:77]: Yes, that's correct.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 3 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:128]: Each channel implementation needs to make sure that their functions are idempotent.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 4 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:132]: It should be called by some higher-level channel reader.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 5 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:27]: Hmm I believe elsewhere we use KB instead of kB.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 6 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:69]: No worries, I can take care of that here.

File: python/ray/dag/tests/experimental/test_mocked_nccl_dag.py (hunk 7 above)
Comment (stephanie-wang) [reply python/ray/dag/tests/experimental/test_mocked_nccl_dag.py:374]: Sorry, I didn't understand what you're suggesting we check?

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 8 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:213]: They can be received in either order. Is that what you're asking?

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 9 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:273]: Should be okay, I'll update the code to just hold the reference instead of calling end_read.

**Comment by anyscalesam** [comment]
discussed with @rkooo567  - changing to p1 right now as it's a general perf optimization. will tackle later.

**Comment by rkooo567** [comment]
required for multi node pp training

File: python/ray/dag/tests/experimental/test_mocked_nccl_dag.py [hunk 11]
```diff
@@ -0,0 +1,355 @@
...
+
+    ray.get([sender.start_mock.remote(), receiver.start_mock.remote()])
+
+    shape = (10,)
+    dtype = torch.float16
+
+    # Test torch.Tensor sent between actors.
+    with InputNode() as inp:
+        dag = sender.send.bind(
+            inp.shape, inp.dtype, inp[0], send_as_dict=inp.send_as_dict
```
Comment (woshiyyya) [inline python/ray/dag/tests/experimental/test_mocked_nccl_dag.py:90]: One question about the output schema, what type of outputs format does our auto infer feature support?IIUC, the actor method can only return one value, the type should be either "a single tensor" or "a dict of tensors"? 

Does that contradict with [this PR](https://github.com/ray-project/ray/pull/47024)? where ADAG can create multiple channels for multiple return values

Comment (stephanie-wang) [reply python/ray/dag/tests/experimental/test_mocked_nccl_dag.py:90]: Great question.

The type can be anything, although types with simple serialization methods are best since this PR requires the user to guarantee serialization order of tensors.

This PR should be compatible with #47024. #47024 works as long as:
1. the actor method is annotated with `num_returns=N`
2. the actor method returns a tuple of N results.
If (2) does not match (1), then an error is thrown.

This PR applies to individual results. So for this kind of function:

```python
def return_two(self):
  return torch.tensor(...), {
    "a": torch.tensor(...),
    "b": torch.tensor(...),
  }
```

if there is no `num_returns` annotation, then the actor returns one value and the schema would be `Tuple[Tensor, Dict[str, Tensor]]`. If `num_returns=2`, then the actor will send on two different channels, one with schema `Tensor` and the other with schema `Dict[str, Tensor]`.

File: python/ray/experimental/channel/intra_process_channel.py [hunk 12]
```diff
@@ -57,7 +57,8 @@ def write(self, value: Any, timeout: Optional[float] = None):
         ctx = ChannelContext.get_current().serialization_context
         ctx.set_data(self._channel_id, value, self._num_readers)
 
-    def read(self, timeout: Optional[float] = None) -> Any:
+    def read(self, timeout: Optional[float] = None, deserialize: bool = True) -> Any:
+        assert deserialize, "Data passed from the actor to itself is never serialized"
```
Comment (rkooo567) [inline python/ray/experimental/channel/intra_process_channel.py:61]: isn't it supposed to be False? `assert not deserialize`?

File: python/ray/experimental/channel/serialization_context.py [hunk 13]
```diff
@@ -8,7 +8,14 @@
 class _SerializationContext:
     def __init__(self):
         self.use_external_transport: bool = False
```
Comment (rkooo567) [inline python/ray/experimental/channel/serialization_context.py:10]:
```suggestion
        self._use_external_transport: bool = False
```

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 14]
```diff
@@ -30,15 +27,26 @@
...
     def __init__(
         self,
         writer: ray.actor.ActorHandle,
         reader_and_node_list: List[Tuple["ray.actor.ActorHandle", str]],
-        gpu_data_typ: "TorchTensorType",
-        cpu_data_typ: Optional["SharedMemoryType"] = None,
-        _gpu_data_channel: Optional["TorchTensorNcclChannel"] = None,
-        _cpu_data_channel: Optional["Channel"] = None,
+        tensor_data_channel: "_TorchTensorNcclChannel",
+        tensor_schema_channel: "Channel",
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:48]: this can include data other than schema right? Like other cpu data structure. Why don't we just keep calling it _cpu_data_channel?

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 15]
```diff
@@ -303,70 +388,92 @@ def __reduce__(self):
...
+        import torch
+
+        for tensor in tensors:
+            assert isinstance(
+                tensor, torch.Tensor
+            ), f"{tensor} must be instance of torch.Tensor"
+
+        # Send the tensors metadata so that the receiver knows what buffers to
+        # allocate.
+        metadata = self._get_send_tensors_metadata(tensors)
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:474]: should we check if the metadata and tensor matches? Or is it checked in the reader side?

**Review by rkooo567 — COMMENTED** [review]
static_shape part looks pretty good to me! 

I'd like to suggest the PR to split to 2 parts;

- support static_shape
- support static_tensor_schema

Before proceeding with `static_tensor_schema`, 
- It is very prone to error. It is common user mixes up cpu and gpu data to send (e.g., vLLM), and it doesn't work at all for that case. 
- Even when it only contains tensors, users should guarantee ordering, and there's no good way to check ordering

Given two these two limitation, I wonder if we should support very limited set of return type? I.e., dict and list. And no other data type other than tensor in the container (so it should be Dict[Any, Tensor] and List[Tensor])

File: python/ray/dag/tests/experimental/test_mocked_nccl_dag.py [hunk 16]
```diff
@@ -0,0 +1,355 @@
...
+            "num_gpus": 2,
+            "num_nodes": 1,
+        }
+    ],
+    indirect=True,
+)
+def test_p2p_static_tensor_schema_error(ray_start_cluster):
+    """
+    Test that when static_tensor_schema=True, an error is thrown when a
+    different number of tensors is found.
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply python/ray/dag/tests/experimental/test_mocked_nccl_dag.py]: This is tested in test_p2p_static_shape_error

File: python/ray/dag/tests/experimental/test_mocked_nccl_dag.py [hunk 17]
```diff
@@ -0,0 +1,355 @@
...
+    # Test sending a dictionary with a different number of entries.
+    spec = [
+        ("a", 1, shape, dtype),
+        ("b", 1, shape, dtype),
+    ]
+    ref = compiled_dag.execute(spec)
+    with pytest.raises(OSError, match="Channel closed"):
+        ray.get(ref)
+
+    # Sending correct shape still errors because the DAG has already been torn
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply python/ray/dag/tests/experimental/test_mocked_nccl_dag.py]: Yeah it's a limitation of this technique :( not sure if there is a way for NCCL to recover from the bad recv.

File: python/ray/dag/tests/experimental/test_mocked_nccl_dag.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply python/ray/dag/tests/experimental/test_mocked_nccl_dag.py]: Sending different dtypes and shapes is actually okay, as long as static_shape=False.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 15 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:440]: It is already checked on the writer in `_get_send_tensors_metadata`

File: python/ray/dag/tests/experimental/test_mocked_nccl_dag.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply python/ray/dag/tests/experimental/test_mocked_nccl_dag.py]: Actually this could probably be improved by storing the exception returned by the actor task as the DAG result and then restarting the DAG. But too complicated for now, will leave a TODO.

[the remark this answers is not in the document]
Comment (rkooo567) [reply python/ray/dag/tests/experimental/test_mocked_nccl_dag.py]: can you add docstring to this argument that any failure can cause dag to be teardown>?

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 18]
```diff
@@ -51,38 +59,28 @@ def __init__(
...
-            writer: The actor that may write to the channel. None signifies the driver.
+            writer: The actor that may write to the channel. None signifies the
+                driver.
             reader_and_node_list: A list of tuples, where each tuple contains a reader
                 actor handle and the node ID where the actor is located.
-            gpu_data_typ: Type information about the GPU tensors
-            cpu_data_typ: Type information about the CPU data
+            tensor_data_channel: A GPU-GPU channel for sending tensor data. Its
+                writer and readers should match the given writer and readers.
+            non_tensor_data_channel: A shared-memory channel for sending
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:68]: I think this is technically wrong? I think cpu tensors are still sent through this channel.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 19]
```diff
@@ -146,58 +135,149 @@ def write(self, value: Any, timeout: Optional[float] = None) -> None:
...
+
+    def write(self, value: Any, timeout: Optional[float] = None) -> None:
+        """
+        Send a value that may contain torch.Tensors that should be sent via
+        external transport.
+
+        This method:
+        1) Serializes `value`. During serialization, all torch.Tensors that are
+        on the default device are extracted and replaced with a unique
+        placeholder. Thus, the serialized value will contain all non-tensor
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:158]: I think this is wrong when the tensor is not on the default device (e.g., cpu tensor on gpu actor)

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 20]
```diff
@@ -146,58 +135,149 @@ def write(self, value: Any, timeout: Optional[float] = None) -> None:
...
+        3) Sends the non-tensor data via the non-tensor data channel.
+
+        If static_non_tensor_data=True was specified, then we only perform step
+        (3) on the first `write` call. The reader is expected to reuse the sent
+        data for subsequent messages.
+        """
+        if isinstance(value, ray.exceptions.RayTaskError):
+            if self._typ.static_shape or self._typ.direct_return:
+                # Raise a fatal error to teardown the DAG.
+                # TODO(swang): Write exceptions to the tensor metadata or
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:171]: Can you add the DAG is permantly dead in the error message and make sure from the test that it prints them?

File: python/ray/experimental/channel/torch_tensor_type.py [hunk 21]
```diff
@@ -124,57 +122,45 @@ def create_channel(
...
                 reader_and_node_list,
                 self,
-                _torch_tensor_allocator=_torch_tensor_allocator,
+                _meta_channel=_tensor_metadata_channel,
             )
 
-        # Transfer via host memory using a shared-memory channel.
-        import torch
+            if _non_tensor_data_channel is None and not self._direct_return:
+                # Create a CPU channel to send non-tensor data.
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_type.py:144]: I think we should call it cpu_channel given it can contain tensors!

File: python/ray/experimental/channel/torch_tensor_type.py [hunk 22]
```diff
@@ -124,57 +122,45 @@ def create_channel(
...
-            num_elements *= dim
-        element_size_bytes = TORCH_DTYPE_ITEMSIZE_MAP[self._dtype]
-        buffer_size_bytes = int(num_elements * element_size_bytes)
-        buffer_size_bytes += TENSOR_METADATA_SIZE_BYTES
-
-        return Channel(writer, reader_and_node_list, buffer_size_bytes)
+        # Data does not require NCCL. Transfer via host memory using a
+        # shared-memory channel.
+        # TODO(swang): Allow the initial max buffer size to bereaders overridden.
+        typ = SharedMemoryType()
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_type.py:161]: this seems unused? `typ`

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 23]
```diff
@@ -146,58 +135,149 @@ def write(self, value: Any, timeout: Optional[float] = None) -> None:
...
+        1) Receives torch.Tensors via the tensor data channel (e.g., NCCL).
+        2) Reads the serialized non-tensor data.
+        3) Deserializes the non-tensor data. During deserialization, replaces
+        all found placeholders with the received torch.Tensors.
 
-        self.serialization_ctx.reset_tensors(tensors)
-        data = self._cpu_data_channel.read()
-        self.serialization_ctx.reset_tensors([])
+        If _direct_return=True was specified, then we skip step (2) and (3) and
+        directly return the data received in (1).
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:234]: can you update the docstring the timeout is not applied on nccl recv?

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 24]
```diff
@@ -146,58 +135,149 @@ def write(self, value: Any, timeout: Optional[float] = None) -> None:
...
+            import torch
+
+            if not isinstance(value, torch.Tensor):
+                # TODO(swang): These errors are currently fatal for the DAG
+                # because there is no way for the receiver to receive the
+                # exception. This could be improved by sending the exception
+                # through the tensor_data_channel's CPU-based metadata channel,
+                # if one exists.
+                raise ValueError("Task annotated with _direct_return=True must "
+                        f"return a CUDA torch.Tensor, instead found value `{value}`")
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:188]: can you mentino the failure is fatal to error messages?

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py [hunk 25]
```diff
@@ -146,58 +135,149 @@ def write(self, value: Any, timeout: Optional[float] = None) -> None:
...
+            elif not value.is_cuda:
+                raise ValueError("Task annotated with _direct_return=True must "
```
Comment (rkooo567) [inline python/ray/experimental/channel/torch_tensor_nccl_channel.py:190]: Is this a temporary semantic?

**Review by rkooo567 — APPROVED** [review]
Generally looks good. I think since the cpu channel can still contain cpu data, we should just name it cpu / gpu_tensor channel!

File: python/ray/experimental/channel/torch_tensor_type.py (hunk 22 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_type.py:161]: It's used in the next line.

File: python/ray/experimental/channel/torch_tensor_nccl_channel.py (hunk 25 above)
Comment (stephanie-wang) [reply python/ray/experimental/channel/torch_tensor_nccl_channel.py:190]: What do you mean?

File: python/setup.py [hunk 26]
```diff
@@ -772,51 +772,55 @@ def has_ext_modules(self):
...
-    include_package_data=True,
-    exclude_package_data={
-        # Empty string means "any package".
-        # Therefore, exclude BUILD from every package:
-        "": ["BUILD"],
-    },
-    zip_safe=False,
-    license="Apache 2.0",
-) if __name__ == "__main__" else None
+(
```
Comment (aslonnie) [inline python/setup.py:775]: is this just an indentation change? why this change is required?
